src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -2223,54 +2223,55 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
       i += 1
     }
 
-    val newCellType = TStruct(
-      "proband" -> genotypeSignature,
-      "father" -> genotypeSignature,
-      "mother" -> genotypeSignature
+    val gSig = genotypeSignature
+
+    val newEntryType = TStruct(
+      "proband" -> gSig,
+      "father" -> gSig,
+      "mother" -> gSig
     )
-    val gsig = genotypeSignature
 
     val oldRowType = rowType
     val newRowType = TStruct(Array(
       rowType.fields(0),
       rowType.fields(1),
       rowType.fields(2),
-      Field(rowType.fields(3).name, TArray(newCellType), 3)
+      Field(rowType.fields(3).name, TArray(newEntryType), 3)
     ))
 
+    val oldGsType = rowType.fieldType(3).asInstanceOf[TArray]
+
     val newRDD = rdd2.mapPartitionsPreservesPartitioning { it =>
       it.map { r =>
         val rvb = new RegionValueBuilder(MemoryBuffer())
         rvb.start(newRowType)
         rvb.startStruct()
 
-        // FIXME: The next four lines allocate
-        val oldRow = UnsafeRow.readStruct(oldRowType, r.region, r.offset)
-        rvb.addAnnotation(newRowType.fields(0).typ, oldRow.get(0))
-        rvb.addAnnotation(newRowType.fields(1).typ, oldRow.get(1))
-        rvb.addAnnotation(newRowType.fields(2).typ, oldRow.get(2))
+        rvb.addField(oldRowType, r, 0)
+        rvb.addField(oldRowType, r, 1)
+        rvb.addField(oldRowType, r, 2)
 
-        val arr = oldRow.getAs[IndexedSeq[Annotation]](3)
         rvb.startArray(nTrios)
+        val gsOffset = oldRowType.loadField(r, 3)
 
         var i = 0
         while (i < nTrios) {
           rvb.startStruct()
 
           // append kid element
-          rvb.addAnnotation(gsig, arr(kidIndices(i)))
+          rvb.addElement(oldGsType, r.region, gsOffset, kidIndices(i))
```

Comment:
Should these be addField rather than addElement?

Reply:
no, the field/element refers to the source, not the destination

Reply:
from what I can tell.

Reply:
Yes, `addElement` adds the `i`th element from an array region value, and `addField` adds a the `i`th field from a struct region value.